### PR TITLE
Added support for remote unix connection to libvirt.

### PR DESCRIPTION
**Pull request by kamilvarga**
Second password for additional verification methods(SASL) added to
libvirt.py, new parameter added to unixconnection, so it can be used
for communication and setup on remote host.

Used for a new test, results availble at: https://github.com/autotest/tp-libvirt/pull/3425

Signed-off-by: Kamil Varga <kvarga@redhat.com>

### CODE REVIEW

File: virttest/utils_conn.py [hunk 1]
```diff
@@ -1817,22 +1829,46 @@ def conn_recover(self):
...
+                    runner("systemctl unmask libvirtd-tcp.socket")
+                    runner("systemctl unmask libvirtd-tls.socket")
                 else:
-                    process.run("systemctl daemon-reload", shell=True)
-                    process.run("systemctl stop libvirtd.socket", shell=True)
-                libvirtd_service.stop()
-                process.run("systemctl start libvirtd.socket", shell=True)
+                    runner("systemctl daemon-reload")
+                    runner("systemctl stop libvirtd.socket")
+                # libvirtd_service.stop()
```
Comment (dzhengfy) [inline virttest/utils_conn.py:1865]: remove this line

File: virttest/utils_conn.py [hunk 2]
```diff
@@ -1817,22 +1829,46 @@ def conn_recover(self):
...
+                runner("systemctl stop libvirtd.service")
+                # runner("systemctl daemon-reload")
```
Comment (dzhengfy) [inline virttest/utils_conn.py:1867]: remove this line

File: virttest/utils_conn.py [hunk 3]
```diff
@@ -1858,19 +1894,25 @@ def conn_setup(self):
...
+        remote_session = None
+        server_ip = self.server_ip
+        server_user = self.server_user
+        server_pwd = self.server_pwd
         sasl_allowed_username_list = self.sasl_allowed_username_list
         traditional_mode = self.traditional_mode
 
         # edit the /etc/libvirt/libvirtd.conf to add auth_unix_ro arg
-        if auth_unix_ro:
+        if auth_unix_ro != 'none':
```
Comment (dzhengfy) [inline virttest/utils_conn.py:1909]: What is the reason for this change? 
If auth_unix_ro == 'none', we should still do the replacement because the default value with '#' in the beginning is polkit, not 'none'.

File: virttest/utils_conn.py [hunk 4]
```diff
@@ -1858,19 +1894,25 @@ def conn_setup(self):
...
             pattern_to_repl = {r".*auth_unix_ro\s*=.*":
                                'auth_unix_ro="%s"' % auth_unix_ro}
             self.client_libvirtdconf.sub_else_add(pattern_to_repl)
 
         # edit the /etc/libvirt/libvirtd.conf to add auth_unix_rw arg
-        if auth_unix_rw:
+        if auth_unix_rw != 'none':
```
Comment (dzhengfy) [inline virttest/utils_conn.py:1915]: same as above

File: virttest/utils_conn.py [hunk 5]
```diff
@@ -1898,19 +1940,19 @@ def conn_setup(self):
...
 
-        if auth_unix_rw == 'sasl':
+        if auth_unix_rw == 'sasl' and sasl_type != 'plain':
             pattern_to_repl = {r".*access_drivers\s*=.*":
                                '#access_drivers="%s"' % access_drivers}
             self.client_libvirtdconf.sub_else_add(pattern_to_repl)
 
         # edit the /etc/libvirt/libvirtd.conf to add unix_sock_ro_perms arg
-        if unix_sock_ro_perms:
+        if unix_sock_ro_perms != '0777':
```
Comment (dzhengfy) [inline virttest/utils_conn.py:1949]: What is the reason for this change?

File: virttest/utils_conn.py [hunk 6]
```diff
@@ -1898,19 +1940,19 @@ def conn_setup(self):
...
             pattern_to_repl = {r".*unix_sock_ro_perms\s*=.*":
                                'unix_sock_ro_perms="%s"' % unix_sock_ro_perms}
             self.client_libvirtdconf.sub_else_add(pattern_to_repl)
 
         # edit the /etc/libvirt/libvirtd.conf to add unix_sock_rw_perms arg
-        if unix_sock_rw_perms:
+        if unix_sock_rw_perms != '0770':
```
Comment (dzhengfy) [inline virttest/utils_conn.py:1955]: What is the reason for this change?

File: virttest/utils_conn.py [hunk 7]
```diff
@@ -1941,41 +1989,68 @@ def conn_setup(self):
             pattern_to_repl = {r".*sasl_allowed_username_list\s*=.*":
                                'sasl_allowed_username_list=%s' % sasl_allowed_username_list}
             self.client_libvirtdconf.sub_else_add(pattern_to_repl)
-        else:
+        elif not self.sasl_type == 'plain':
```
Comment (dzhengfy) [inline virttest/utils_conn.py:1992]: This is not recommended coding style in python.  
`elif self.sasl_type != 'plain':` is better

File: virttest/utils_conn.py [hunk 8]
```diff
@@ -1941,41 +1989,68 @@ def conn_setup(self):
...
         self.client_saslconf.sub_else_add(pattern_to_repl)
 
         # restart libvirtd service on server
         if restart_libvirtd == "yes":
             try:
+                if not self.run_on_remote:
+                    current_session = client_session
+                    runner = process.run
+                else:
+                    current_session = remote.wait_for_login('ssh',
```
Comment (dzhengfy) [inline virttest/utils_conn.py:2012]: Could we use `current_session = remote_session` ? Any reason to create a separate session for current_session?

File: virttest/utils_conn.py [hunk 9]
```diff
@@ -1941,41 +1989,68 @@ def conn_setup(self):
...
+                        runner("systemctl stop libvirtd-tcp.socket")
+                        runner("systemctl stop libvirtd-tls.socket")
                         libvirtd_service.stop()
-                        process.run("systemctl mask libvirtd.socket", shell=True)
-                        process.run("systemctl mask libvirtd-admin.socket", shell=True)
-                        process.run("systemctl mask libvirtd-ro.socket", shell=True)
-                        process.run("systemctl mask libvirtd-tcp.socket", shell=True)
-                        process.run("systemctl mask libvirtd-tls.socket", shell=True)
-                        process.run("systemctl daemon-reload", shell=True)
+                        runner("systemctl mask libvirtd.socket")
```
Comment (dzhengfy) [inline virttest/utils_conn.py:2038]: Is there any potential risks to remove 'shell=True' for process.run?

File: virttest/utils_conn.py (hunk 3 above)
Comment (kamilvarga) [reply virttest/utils_conn.py:1909]: Unfortunately, the dictionary is initialized to 'none' at line 1747 if there is no any other value provided to method. That means, the value of auth_unix_ro was always set to none, when no other value was provided. Same case was for auth_unix_rw and similar to unix_sock_ro_perms, unix_sock_rw_perms. Those changes are not intended by test and therefore when no value is provided to UnixConnection class, the default values in libvirtd.conf or libvirt.conf config files are used.

File: virttest/utils_conn.py (hunk 4 above)
Comment (kamilvarga) [reply virttest/utils_conn.py:1915]: Please see my explanation above.

File: virttest/utils_conn.py (hunk 5 above)
Comment (kamilvarga) [reply virttest/utils_conn.py:1949]: Please see my explanation above.

File: virttest/utils_conn.py (hunk 6 above)
Comment (kamilvarga) [reply virttest/utils_conn.py:1955]: Please see my explanation above.

File: virttest/utils_conn.py (hunk 7 above)
Comment (kamilvarga) [reply virttest/utils_conn.py:1992]: Will fix in next commit.

File: virttest/utils_conn.py (hunk 1 above)
Comment (kamilvarga) [reply virttest/utils_conn.py:1865]: Will fix in next commit.

File: virttest/utils_conn.py (hunk 2 above)
Comment (kamilvarga) [reply virttest/utils_conn.py:1867]: Will fix in next commit.

File: virttest/utils_conn.py (hunk 8 above)
Comment (kamilvarga) [reply virttest/utils_conn.py:2012]: I found, the session might be interfere, when the libvirtd services are restarted and therefore there are two independent sessions.

File: virttest/utils_conn.py (hunk 9 above)
Comment (kamilvarga) [reply virttest/utils_conn.py:2038]: I didn't found any. The shell = True will just create a subshell during the execution and it is not supported by the remote runner, so I didn't found a reason to have it here. Are you aware of any possible issue?

File: virttest/utils_conn.py (hunk 1 above)
Comment (kamilvarga) [reply virttest/utils_conn.py]: Updated as suggested.

File: virttest/utils_conn.py (hunk 2 above)
Comment (kamilvarga) [reply virttest/utils_conn.py]: Updated as suggested.

File: virttest/utils_conn.py (hunk 7 above)
Comment (kamilvarga) [reply virttest/utils_conn.py]: Updated as suggested.

File: virttest/utils_conn.py (hunk 9 above)
Comment (dzhengfy) [reply virttest/utils_conn.py:2036]: I remember process.run might get certain unexpected error without shell=True sometimes. But if you have run all the related test cases, I am fine with it so far.

**Review by dzhengfy — APPROVED** [review]
LGTM

**Comment by dzhengfy** [comment]
@Yingshun Could you have a review?

File: virttest/utils_conn.py (hunk 8 above)
Comment (Yingshun) [reply virttest/utils_conn.py:2010]: @kamilvarga Sorry, server_session works on my environment, could you please explain some details about the problem you faced? The session is RemoteSession(ShellSession) object, I think it should not be affected by libvirtd service restarting. This looks weird~
And below are my testing:
```
>>> conn_obj=utils_conn.UNIXConnection(xxx)
>>> conn_obj.client_session.cmd("hostname")
'dell-per740-41.lab.eng.pek2.redhat.com\n'
>>> conn_obj.server_session.cmd("hostname")
'dell-per740-43.lab.eng.pek2.redhat.com\n'
>>> from virttest import utils_libvirtd
>>> server_libvirtd=utils_libvirtd.Libvirtd(session=conn_obj.server_session)
>>> server_libvirtd.is_running()
False
>>> server_libvirtd.restart()
True
>>> server_libvirtd.is_running()
True
>>> conn_obj.server_session.cmd("hostname")
'dell-per740-43.lab.eng.pek2.redhat.com\n'
>>> conn_obj.client_session.cmd("hostname")
'dell-per740-41.lab.eng.pek2.redhat.com\n'

```

File: virttest/utils_conn.py (hunk 3 above)
Comment (Yingshun) [reply virttest/utils_conn.py:1907]: I wonder if we can update the default value of these settings to None or just remove the default values?  I think it might be an issue if we need to set auth_unix_ro to 'none' in some cases... What do you think？ @kamilvarga @dzhengfy

File: virttest/utils_conn.py (hunk 8 above)
Comment (kamilvarga) [reply virttest/utils_conn.py:2010]: Will check again with full script.

File: virttest/utils_conn.py (hunk 3 above)
Comment (dzhengfy) [reply virttest/utils_conn.py:1907]: @kamilvarga, after discussion with @Yingshun , we think  'none' is a valid value for those parameters, so below default value 'none' should be removed.  ` init_dict['auth_unix_rw'] = init_dict.get('auth_unix_rw', 'none')` 
Then there will not be a 'none' value for those parameters. So your change for L1909` if auth_unix_ro != 'none':` can be removed.

Comment (kamilvarga) [reply virttest/utils_conn.py:1907]: Will check.

File: virttest/utils_conn.py [hunk 10]
```diff
@@ -1762,9 +1763,20 @@ def __init__(self, *args, **dargs):
             'sasl_allowed_username_list', '["root/admin" ]')
         init_dict['traditional_mode'] = init_dict.get(
             'traditional_mode', 'no')
+        init_dict['run_on_remote'] = init_dict.get('run_on_remote', 'no')
 
         super(UNIXConnection, self).__init__(init_dict)
 
+        if init_dict['run_on_remote'] == 'yes':
+            self.client_ip = self.server_ip
```
Comment (chunfuwen) [inline virttest/utils_conn.py:1771]: Do we really need override client_ip value under run_on_remote mode ? it looks like later logic still use self.server_ip to do relevant operations.

File: virttest/utils_conn.py (hunk 3 above)
Comment (kamilvarga) [reply virttest/utils_conn.py]: Updated as suggested.

File: virttest/utils_conn.py (hunk 5 above)
Comment (kamilvarga) [reply virttest/utils_conn.py]: Updated per discussion above.

File: virttest/utils_conn.py (hunk 6 above)
Comment (kamilvarga) [reply virttest/utils_conn.py]: Updated per discussion above.

File: virttest/utils_conn.py (hunk 8 above)
Comment (kamilvarga) [reply virttest/utils_conn.py]: Reworked and tested with the following results:
<pre><font color="#2A7BDE">JOB ID     : 20f8df233d8e53671d985160efb74b3227b64c5b</font>
<font color="#2A7BDE">JOB LOG    : /root/avocado/job-results/job-2021-06-01T18.03-20f8df2/job.log</font>
 (01/17) type_specific.io-github-autotest-libvirt.virsh.remote_with_unix.positive_testing.default_unix_config: <font color="#33DA7A">PASS</font> (58.55 s)
 (02/17) type_specific.io-github-autotest-libvirt.virsh.remote_with_unix.positive_testing.readonly: <font color="#33DA7A">PASS</font> (57.84 s)
 (03/17) type_specific.io-github-autotest-libvirt.virsh.remote_with_unix.positive_testing.customized_unix_sock_dir: <font color="#33DA7A">PASS</font> (59.84 s)
 (04/17) type_specific.io-github-autotest-libvirt.virsh.remote_with_unix.positive_testing.default_config_and_start_vm: <font color="#33DA7A">PASS</font> (59.98 s)
 (05/17) type_specific.io-github-autotest-libvirt.virsh.remote_with_unix.positive_testing.allow_sasl_plain: <font color="#33DA7A">PASS</font> (72.84 s)
 (06/17) type_specific.io-github-autotest-libvirt.virsh.remote_with_unix.positive_testing.allow_sasl_krb_user: <font color="#33DA7A">PASS</font> (101.98 s)
 (07/17) type_specific.io-github-autotest-libvirt.virsh.remote_with_unix.positive_testing.allow_sasl_users: <font color="#33DA7A">PASS</font> (70.86 s)
 (08/17) type_specific.io-github-autotest-libvirt.virsh.remote_with_unix.positive_testing.socket_access_controls: <font color="#33DA7A">PASS</font> (70.60 s)
 (09/17) type_specific.io-github-autotest-libvirt.virsh.remote_with_unix.positive_testing.socket_with_auth_conf: <font color="#33DA7A">PASS</font> (56.44 s)
 (10/17) type_specific.io-github-autotest-libvirt.virsh.remote_with_unix.positive_testing.socket_with_polkit_and_acl_control: <font color="#33DA7A">PASS</font> (63.93 s)
 (11/17) type_specific.io-github-autotest-libvirt.virsh.remote_with_unix.positive_testing.socket_with_polkit_rw: <font color="#33DA7A">PASS</font> (63.77 s)
 (12/17) type_specific.io-github-autotest-libvirt.virsh.remote_with_unix.negative_testing.no_allowed_sasl_user: <font color="#33DA7A">PASS</font> (114.74 s)
 (13/17) type_specific.io-github-autotest-libvirt.virsh.remote_with_unix.negative_testing.allow_sasl_user_with_readonly: <font color="#33DA7A">PASS</font> (113.20 s)
 (14/17) type_specific.io-github-autotest-libvirt.virsh.remote_with_unix.negative_testing.readonly_without_auth: <font color="#33DA7A">PASS</font> (59.84 s)
 (15/17) type_specific.io-github-autotest-libvirt.virsh.remote_with_unix.negative_testing.socket_access_controls.cfg_file: <font color="#33DA7A">PASS</font> (68.87 s)
 (16/17) type_specific.io-github-autotest-libvirt.virsh.remote_with_unix.negative_testing.socket_access_controls.socket_file: <font color="#33DA7A">PASS</font> (61.93 s)
 (17/17) type_specific.io-github-autotest-libvirt.virsh.remote_with_unix.negative_testing.socket_with_polkit_ro: <font color="#33DA7A">PASS</font> (62.51 s)
<font color="#2A7BDE">RESULTS    : PASS 17 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0</font>
<font color="#2A7BDE">JOB TIME   : 1220.71 s</font></pre>

File: virttest/utils_conn.py (hunk 10 above)
Comment (kamilvarga) [reply virttest/utils_conn.py]: This part was reworked in last update.

**Comment by kamilvarga** [comment]
@dzhengfy , @Yingshun, @chunfuwen  The unix connection was updated per your comments and suggestions. Please take a look.

File: virttest/utils_conn.py [hunk 11]
```diff
@@ -1815,32 +1832,74 @@ def conn_recover(self):
...
             else:
                 libvirtd_service.restart()
+            current_session.close()
         except (remote.LoginError, aexpect.ShellError,
                 process.CmdError) as detail:
             raise ConnServerRestartError(detail)
 
         logging.debug("UNIX connection recover successfully.")
 
+    def session_creator(self):
```
Comment (Yingshun) [inline virttest/utils_conn.py:1868]: I wonder if it's possible to create sessions via `_new_client_session()/_new_server_session()`. It might look simpler... What do you think?

File: virttest/utils_conn.py [hunk 12]
```diff
@@ -1939,45 +2005,59 @@ def conn_setup(self):
                                }
             self.client_hosts.sub_else_add(pattern_to_repl)
             pattern_to_repl = {r".*sasl_allowed_username_list\s*=.*":
-                               'sasl_allowed_username_list=%s' % sasl_allowed_username_list}
+                               'sasl_allowed_username_list=%s'
+                               % sasl_allowed_username_list}
             self.client_libvirtdconf.sub_else_add(pattern_to_repl)
-        else:
+        elif self.sasl_type != 'plain':
```
Comment (Yingshun) [inline virttest/utils_conn.py:2011]: Do we need to cover 'else' here as in the previous code?

Comment (kamilvarga) [reply virttest/utils_conn.py:2011]: Else statement from here is covered on line 2017.

**Review by chunfuwen — APPROVED** [review]
lgtrm

File: virttest/utils_conn.py [hunk 13]
```diff
@@ -1815,32 +1832,74 @@ def conn_recover(self):
...
+        Prepare a local/remote sessions as required and return them back.
+        """
+        server_ip = self.server_ip
+        client_ip = self.client_ip
+        server_user = self.server_user
+        client_user = self.client_user
+        server_pwd = self.server_pwd
+        client_pwd = self.client_pwd
+        if self.run_on_remote:
+            client_session = remote.wait_for_login('ssh',
```
Comment (dzhengfy) [inline virttest/utils_conn.py:1879]: I think it looks like a duplicated sessio. Could you help to check if get_client_session() and get_server_session() meet your need?

Comment (kamilvarga) [reply virttest/utils_conn.py:1879]: @dzhengfy It works on RHEL8.4, but I had issues on RHEL9.0, unfortunately cannot reproduce due to latest RHEL9.0 issues. What do you suggest? Should I keep try it on RHEL9.0 or update per working steps from 8.4?

Comment (kamilvarga) [reply virttest/utils_conn.py]: Updated as discussed, using class variables - self.client_session and self.server_session. It works on RHEL8.4 as shown below. Cannot verify on RHEL9.0 as there is known bug with SASL authentification at a moment.
Results:
<pre>avocado run --vt-type libvirt --vt-machine-type q35 virsh.remote_with_unix.positive_testing.allow_sasl_plain
WARNING:root:No python imaging library installed. Screendump and Windows guest BSOD detection are disabled. In order to enable it, please install python-imaging or the equivalent for your distro.
WARNING:root:No python imaging library installed. PPM image conversion to JPEG disabled. In order to enable it, please install python-imaging or the equivalent for your distro.
<font color="#2A7BDE">JOB ID     : 2284f8e707c0c4f12f99e0ded16f9f59504bc517</font>
<font color="#2A7BDE">JOB LOG    : /root/avocado/job-results/job-2021-06-16T09.08-2284f8e/job.log</font>
 (1/1) type_specific.io-github-autotest-libvirt.virsh.remote_with_unix.positive_testing.allow_sasl_plain: <font color="#33DA7A">PASS</font> (70.33 s)
<font color="#2A7BDE">RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0</font>
<font color="#2A7BDE">JOB TIME   : 71.67 s</font>
</pre>

File: virttest/utils_conn.py (hunk 11 above)
Comment (kamilvarga) [reply virttest/utils_conn.py:1868]: Updated by class variables self.client_session and self.server_session as it works on RHEL8.4 as mentioned below.

**Review by dzhengfy — APPROVED** [review]
LGTM

**Review by Yingshun — APPROVED** [review]
